tests/test_username_api.py

```diff
@@ -1,35 +1,28 @@
 import json
 import logging
 import os.path
-import random
-import string
 
 from parameterized import parameterized
-import pytest
+import exrex
 import yaml
 
 import username_api
 
 data = yaml.load(open(os.path.join('tests', 'test_data.yml')))
 websites = yaml.load(open('websites.yml'))
 
-invalid_username = '$very%long{invalid}user(name)'
 
-
-def generate_random_username(website, length):
-    first_char_letter = False
-    if website == 'openhub':
-        first_char_letter = True
-    return ''.join(generate_random_character(length,
-                                             first_char_letter))
-
-
-def generate_random_character(length, first_char_letter):
-    for i in range(length):
-        if first_char_letter and i == 0:
-            yield random.choice(string.ascii_letters)
-        else:
-            yield random.choice(string.ascii_letters + string.digits)
+def generate_random_username(website):
+    result = ''
+    username_pattern = websites['username_patterns'][website]
+    while True:
+        random_username = exrex.getone('[{chars}]{{{max}}}'.format(
```

Comment:
also include min please.  I would prefer we have two random usernames, one min, and one max.

Reply:
@jayvdb when I choose length between min and max, they generated username that exist even I ran the test twice. so I set the length to always be max.

Reply:
I think you should look at the `pytest-repeat` type plugins (there are others) which deal with tests which may fail.

Reply:
hmm... I think generating a username that most likely to be invalid is better than repeating test several time. and if we use plugins like `pytest-repeat` how many time we should repeat?

Reply:
"most likely" is very fine.

but it doesnt test that the minimum length is also viable.  separate test for that.  min is mentioned in the issue you are solving.

investigate how many times.  it will obviously be different for min vs max.

it is useless us saying the 'min' is 3 chars if there are no 3 char usernames that are available :P

Reply:
I see. this issue will need some more work...